src/advanced_inv.cpp

```diff
@@ -760,49 +797,37 @@ bool advanced_inventory_pane::is_filtered( const item *it ) const
     return filtercache[ str ];
 }
 
-template <typename Container>
-static itemslice i_stacked( Container items )
+// roll our own, to handle moving stacks better
+typedef std::vector<std::list<item*>> itemstack;
+
+template <typename T>
+static itemstack i_stacked(T items)
 {
     //create a new container for our stacked items
-    itemslice islice;
-
-    //iterate through all items in the vector
-    for( auto &items_it : items ) {
-        if( items_it.count_by_charges() ) {
-            // Those exists as a single item all the item anyway
-            islice.push_back( std::make_pair( &items_it, 1 ) );
-            continue;
-        }
-        bool list_exists = false;
-
-        //iterate through stacked item lists
-        for( auto &elem : islice ) {
-            //check if the ID exists
-            item *first_item = elem.first;
-            if( first_item->type->id == items_it.type->id ) {
-                //we've found the list of items with the same type ID
-
-                if( first_item->stacks_with( items_it ) ) {
-                    //add it to the existing list
-                    elem.second++;
-                    list_exists = true;
-                    break;
-                }
+    itemstack stacks;
+    // used to recall index we stored `itype_id' item at in itemstack
+    std::unordered_map<itype_id, int> cache;
```

Comment:
This doesn't stack correctly. For example this item list (seen in the pickup menu):
- aluminum can of chocolate drink
- aluminum can of cream soda
- aluminum can of chocolate drink

The AIM list shows it item separately even through the two should form one stack. They stack correctly in the inventory.

Cause is probably that you call `stacks_with` only on the first item in the stack. In the example above, the stack first contains the first can (`cache["can"]==0`).

Than the second can is added, tried on `stack[0]` (0 because `cache["can"]` is 0), but it does not stack with the other can, so `got_stacked` is false. The can is added as a separate stack item, but the cache is updated `cache["can"]` is now 1.

The third can is tested against the second can and does not stack either.

The base problem is that the cache only stores the item id, but the stack may contain items with the same id on different places (also think about items of the same type, but with different damage status).